drivers/spi/spi_shell.c

```diff
@@ -4,54 +4,275 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#include <string.h>
 #include <errno.h>
 #include <stdbool.h>
 #include <stdint.h>
 #include <stdlib.h>
 #include <zephyr/drivers/spi.h>
 #include <zephyr/shell/shell.h>
 #include <zephyr/sys/util.h>
+#include <zephyr/devicetree.h>
 
-#define TXRX_ARGV_BYTES     (1)
-#define CONF_ARGV_DEV       (1)
+#define TXRX_ARGV_SPI_DEV (1)
+#define TXRX_ARGV_BYTES   (2)
+
+#define CONF_ARGV_SPI_DEV   (1)
 #define CONF_ARGV_FREQUENCY (2)
 #define CONF_ARGV_SETTINGS  (3)
 
-#define CS_ARGV_GPIO_DEV    (1)
-#define CS_ARGV_GPIO_PIN    (2)
-#define CS_ARGV_GPIO_FLAGS  (3)
+#define CS_ARGV_SPI_DEV    (1)
+#define CS_ARGV_GPIO_DEV   (2)
+#define CS_ARGV_GPIO_PIN   (3)
+#define CS_ARGV_GPIO_FLAGS (4)
 
 /* Maximum bytes we can write and read at once */
 #define MAX_SPI_BYTES MIN((CONFIG_SHELL_ARGC_MAX - TXRX_ARGV_BYTES), 32)
 
-static struct device *spi_device;
-static struct spi_config config = {.frequency = 1000000,
-				   .operation = SPI_OP_MODE_MASTER | SPI_WORD_SET(8)};
+/* Runs the given fn only if the node_id belongs to a spi device, which is on an okay spi bus. */
+#define RUN_FN_ON_SPI_DEVICE(node_id, fn)                                                          \
+	COND_CODE_1(DT_ON_BUS(node_id, spi), \
+	(COND_CODE_1(DT_NODE_HAS_STATUS_OKAY(DT_BUS(node_id)), \
+	(fn), ())), ())
+
+/* Create specified number of empty structs, separated by ',' */
+#define _EMPTY_STRUCT_INST(idx, list) {0}
+#define CREATE_NUM_EMPTY_STRUCTS(num) LISTIFY(num, _EMPTY_STRUCT_INST, (,))
+
+/* Struct representing either a spi bus or a spi device. */
+struct spi_thing {
```

Comment:
maybe `struct spi_shell_device` is better name?

Reply:
I think yes, I replaced all instances of `spi_thing` with `spi_shell_device`.